osu.Game.Rulesets.Osu/Edit/Checks/CheckSpinnerRecoveryTime.cs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using System.Linq;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets.Edit;
+using osu.Game.Rulesets.Edit.Checks.Components;
+using osu.Game.Rulesets.Osu.Objects;
+
+namespace osu.Game.Rulesets.Osu.Edit.Checks
+{
+    public class CheckSpinnerRecoveryTime : ICheck
+    {
+        public CheckMetadata Metadata { get; } = new CheckMetadata(CheckCategory.Compose, "Spinner recovery time");
+
+        public IEnumerable<IssueTemplate> PossibleTemplates => new IssueTemplate[]
+        {
+            new IssueTemplateSpinnerRecoveryTooShort(this)
+        };
+
+        public IEnumerable<Issue> Run(BeatmapVerifierContext context)
+        {
+            double? recommendedRecoveryTime = getRecommendedRecoveryTime(context.InterpretedDifficulty);
+
+            if (!recommendedRecoveryTime.HasValue)
+                yield break;
+
+            foreach (var (firstObject, secondObject) in context.Beatmap.HitObjects.Zip(context.Beatmap.HitObjects.Skip(1)))
+            {
+                if (firstObject is not Spinner spinner)
+                    continue;
+
+                double timeDifference = secondObject.StartTime - spinner.EndTime;
+
+                var timingPoint = context.Beatmap.ControlPointInfo.TimingPointAt(spinner.EndTime);
+
+                double timeDifferenceInBeats = timeDifference / timingPoint.BeatLength;
```

Comment:
So I figured I'd cross-check this against what is arguably the reference implementation of this, namely the MapsetVerifier implementation of this, and noticed it to be [considerably more involved than this](https://github.com/Naxesss/MapsetChecks/blob/8efbf76d17aa495732b2de0b75e33132169cecd4/Checks/standard/spread/CheckSpinnerRecovery.cs#L61-L74).

I asked around and apparently this is because the osu! RC says:

> Overall rules and guidelines apply to every kind of osu! difficulty. Rhythm-related rules and guidelines apply to approximately 180 BPM beatmaps with 4/4 time signatures. If your song is drastically faster or slower, some variables might be different, as detailed in [Scaling BPM on the ranking criteria](https://osu.ppy.sh/wiki/en/Ranking_criteria/Scaling_BPM).

And in that link, more specifically:

> For example, osu! Easy difficulties require 4 beats of recovery time after spinners according to the Ranking Criteria. At 180 BPM, that is true, however at 120 BPM, it would be fine to use 3 beats of recovery, and that could gradually shift to 2 beats at 90 BPM.

So it looks like possibly this isn't as simple as you'd think it was from reading the RC alone. While I'm still not entirely understanding why the MapsetVerifier implementation only uses two reference BPM values rather than the whole 90/120/180/240 gamut (maybe @Naxesss can advise here?), we should probably be mirroring MV here as I believe it to be the de-facto standard in these sorts of RC checks.

Reply:
Ideally, the recovery time would be expressed in milliseconds per difficulty level. That way it is BPM independent. So that's one of the things MV checks against, i.e. 1000 ms, 500 ms, 250 ms for Easy, Normal, and Hard respectively.

However, the RC expresses it in beats to make it easier to check for in the editor. Apparently it also states that
> For the most part, 120 BPM and 240 BPM act as breaking points for when a map's rhythm should be doubled and halved according to guidelines.

[Which results in a BPM-dependent recovery time.](https://github.com/Naxesss/MapsetChecks/blob/8efbf76d17aa495732b2de0b75e33132169cecd4/Checks/standard/spread/CheckSpinnerRecovery.cs#L161C1-L161C1) Compared to the ideal method, this allows shorter recovery times at high BPMs and requires longer ones at low BPMs (iirc).

So basically, we are bound to get false-positives in one direction and false-negatives in the other depending on who you ask. In this case false-positives are more harmful than false-negatives; we'd like issues to be taken seriously. Therefore MV requires both methods to fail before an issue is generated.

Reply:
Oh that's pretty interesting, and I didn't see this check before. I'll port the logic, and update this soon

Reply:
Oh that's pretty interesting, and I didn't see this check before. I'll port the logic, and update this soon